### PR TITLE
Merge retries and error queue into a single recoverability feature

### PR DESCRIPTION
**Pull request by andreasohlund**
Paves the way for #3792 

### TODO

- [x] Merge FLR
- [x] Merge SLR
- [x] [Doco/Upgrade guide PR](https://github.com/Particular/docs.particular.net/pull/1608)

Note that https://github.com/Particular/NServiceBus/pull/3814/commits/0ea4a19f3aa6fa266ac6ac8faeb7ad659e584e64 is needed since we now perform the "does transport support mode x" check earlier in the chain (its a side effect of - GetRequiredTransactionModeForReceives)

### CODE REVIEW

**Comment by andreasohlund** [comment]
@Particular/nservicebus-maintainers @tmasternak @SzymonPobiega please review

File: src/NServiceBus.Core/Recoverability/Recoverability.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        protected internal override void Setup(FeatureConfigurationContext context)
+        {
+            var errorQueue = context.Settings.ErrorQueueAddress();
+            context.Settings.Get<QueueBindings>().BindSending(errorQueue);
+
+            var transportTransactionMode = context.Settings.GetRequiredTransactionModeForReceives();
+            var failureInfoStorage = new FailureInfoStorage(context.Settings.Get<int>(FailureInfoStorageCacheSizeKey));
+            var localAddress = context.Settings.LocalAddress();
+
+            var incomingStepSequence = context.Pipeline.Register("MoveFaultsToErrorQueue", b => new MoveFaultsToErrorQueueBehavior(
```
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: why explicitly defining a step id? (and creating the required registration api?)

File: src/NServiceBus.Core/Recoverability/Recoverability.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                b.Build<CriticalError>(),
+                localAddress,
+                transportTransactionMode,
+                failureInfoStorage), "Moves failing messages to the configured error queue");  //context.Pipeline.Register(new MoveFaultsToErrorQueueBehavior.Registration(context.Settings.LocalAddress(), transportTransactionMode, failureInfoStorage));
+
+            context.Pipeline.Register("AddExceptionHeaders", new AddExceptionHeadersBehavior(), "Adds the exception headers to the message");
```
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: same as above

File: src/NServiceBus.Core/Recoverability/Recoverability.cs (hunk 1 above)
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: edit: nevermind, with this pr we still need the ordering. ignore this.

**Comment by timbussmann** [comment]
why does this PR not also merge or at least align the configuration?
    
    config.DisableFirstLevelRetries();
    config.SecondLevelRetries().Disable();

looks really bad. Can change this to either `config.FirstLevelRetries().Disable()` or merge them to `config.Recoverability().DisableFirstLevelRetries()` etc?

what's the story on the `protected virtual bool DisableFLR()` and `DisableSLR` methods on `DefaultServer`? Why is that needed (and why not properties?)? Not a huge fan of all these subclasses.

**Comment by Scooletz** [comment]
I like the idea of merging these two. 

What @timbussmann :point_up: said. One API should be chosen. Virtual methods to select some features looks leaky - more things can be added.

Beside it, PR in the current shape uses:
- `DelayedRetryPolicy` or `SecondLevelRetryPolicy`
- `ImmediateRetriesEnabled` or `FirstLevelRetryPolicy`.

Using two sets of names is confusing. Either pick new (`Delayed`/`Immediate`) or stick to the previous.

**Comment by timbussmann** [comment]
After playing a bit around with the code I can see how the `protected virtual bool DisableFLR()` got in there: the new API allows to disable SLR/FLR, but since that's the default on the test, there is no config option to "re-enable" the retries at a later test specific configuration. I think this shows that the API may not be ideal, and instead of adding an `Enable()` option (which would allow us to remove the virtual methods and subclasses) why don't we add the following configuration options instead:
* `config.SecondLevelRetries().NumberOfRetries = 0` and `config.FirstLevelRetries().NumberOfRetries = 0` will be equivalent to disabling the feature (for slr unless there is a custom policy configured)
* since we're touching the API, also provide `config.SecondLevelRetries().TimeIncrease` to replicate the XML based config
* the XML config section also has an `Enabled` property which can be translated to `NumberOfRetries = 0`

This means there are no longer mutliple ways of disabling features (configure retries to be zero, disable the feature, or disable on the config section). Possible downside could be that setting retries to zero is less obvious than disabling them altogether from user perspective.

At a later PR we could unify the API to something like `config.Recoverability().ImmediateRetries/DelayedRetries/DelayedTimeIncrease/CustomDelayPolicy`

what do you guys think? @Scooletz @tmasternak @bording @andreasohlund

**Comment by MarcinHoppe** [comment]
@timbussmann When doing this, are you guys taking a look at #3593 and Particular/PlatformDevelopment#756? I recall there was some thinking about API ehancements already.

**Comment by Scooletz** [comment]
@timbussmann I like the idea of having a single way of enabling/disabling features. I don't find intuitive having properties at the end of chain though. I'd go with either `.NumberOfRetries(0)` or `.SetNumberOfRetries(0)`.

**Comment by timbussmann** [comment]
@MarcinHoppe I had these ideas in mind but was wondering how far we should go in this PR (rethink the public API too or keep the public API changes minimal. But admitted, my proposal is also quite a change). My latest proposal (`config.Recoverability().ImmediateRetries/DelayedRetries/DelayedTimeIncrease/CustomDelayPolicy`) seems to align well with #3593 (keeping the custom policy simple for now). if we go with `config.FirstLevelRetries()` we keep the impact lower and can merge them in a later PR at the cost of updating docs twice. Having SLR & FLR separated reduces the risk of confusion around the `TimeIncrease` config (you could write something like `config.Retries().ImmediateRetries(5).DelayRetry(TimeSpan.FromMinutes(1))`)

@Scooletz Seems to make sense since that would allow for better chaining as suggested in #3593 although I don't want to go down the fluent API path (but chaining seems to make sense, e.g. `config.SecondLevelRetries().NumberOfRetries(5).TimeIncrease(TimeSpan.FromMinutes(5));`

**Comment by MarcinHoppe** [comment]
@timbussmann I'm not saying we should do #3593 right now, but if you're changing the API anyways, there's somthing to think about here. Polly is a good source of inspiration.

**Comment by timbussmann** [comment]
a good source of inspiration indeed but we need to be very careful with fluent APIs. That's definitely something I'd keep out of this PR (I think we agree on that). Trying to create a PR based on this changes using my proposal, then we can move the discussion to that PR.

File: src/NServiceBus.Core/Recoverability/Recoverability.cs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            if (retriesConfig != null)
+            {
+                numberOfRetries = retriesConfig.Enabled ? retriesConfig.NumberOfRetries : 0;
+                timeIncrease = retriesConfig.TimeIncrease;
+            }
+
+            return new DefaultSecondLevelRetryPolicy(numberOfRetries, timeIncrease);
+        }
+
+        bool IsDelayedRetriesEnabled(ReadOnlySettings settings)
```
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: the original feature had `DependsOn<DelayedDeliveryFeature>()` which is now `DependsOnOptionally<DelayedDeliveryFeature>();`. It seems like SLR doesn't properly check whether the delayed deliver feature is enabled?

File: src/NServiceBus.Core/Recoverability/Recoverability.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+                context.Pipeline.Register("SecondLevelRetries", b => new SecondLevelRetriesBehavior(retryPolicy, localAddress, failureInfoStorage), "Performs second level retries");
+            }
+
+            if (IsImmediateRetriesEnabled(context.Settings))
+            {
+                var maxRetries = GetMaxRetries(context.Settings);
+                var retryPolicy = new FirstLevelRetryPolicy(maxRetries);
+
+                context.Pipeline.Register("FirstLevelRetries", b => new FirstLevelRetriesBehavior(failureInfoStorage, retryPolicy), "Performs first level retries");
```
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: the slr behavior was registered with `DependencyLifecycle.InstancePerCall` but it seems that there was no real reason for that?

File: src/NServiceBus.Core/Recoverability/Recoverability.cs (hunk 3 above)
Comment (tmasternak) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: Would it be enough to just remove `DependsOnOptionally<DelayedDeliveryFeature>();` ?

Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: DependsOnOptionally also ensures that the dependency is initialized first, so I think we need it (but we could double check that). But it seems we need to code that check into `IsDelayedRetriesEnabled` somehow.

Comment (andreasohlund) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: I just copied it over, pretty sure we need to figure out if the TM is enabled or not. I'll digg deeper

Comment (andreasohlund) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: Found it, the old code used to have dependson 

https://github.com/Particular/NServiceBus/blob/release-6.0.0/src/NServiceBus.Core/Recoverability/SecondLevelRetries/SecondLevelRetries.cs#L17

I changed that to optionally but missed to check if it's enabled or not. Will fix!

Comment (andreasohlund) [inline src/NServiceBus.Core/Recoverability/Recoverability.cs]: After talking to @timbussmann and @tmasternak we concluded that even though this is a bug, the behavior is the same as in v5. @timbussmann will verify this and raise relevant issues.

We'll leave this as is for this PR

**Pull request merged**